File: LeetCode_Python/slidingWindow/nSums/15.3-sum.py
#
# @lc app=leetcode id=15 lang=python3
#
# [15] 3Sum
#

# @lc code=start
class idxPair:
        def __init__(self, idx, val) -> None:
            self.idx = idx
            self.val = val
        def __lt__(self, other):
            return self.val < other.val
        def __gt__(self, other):
            return self.val > other.val
        def __ge__(self, other):
            return self.val >= other.val
        def __le__(self, other):
            return self.val <= other.val
        def __eq__(self, other):
            return self.val == other.val

    

class Solution:
    def sortedTwoSumPairs(self, nums: List[int], start: int, target: int):     
        # find all unique value pairs and save these pairs to an array 
        pLeft = start
        pRight = len(nums)-1
        result = []

        while (pLeft < pRight):
            sumTwo = nums[pLeft] + nums[pRight]
            valLeft = nums[pLeft]
            valRight = nums[pRight]

            if (sumTwo == target):
                result.append([valLeft, valRight])
                while (nums[pLeft]==valLeft and pLeft < pRight):
                    pLeft+=1
                while (nums[pRight]==valRight and pLeft < pRight):
                    pRight-=1    

            elif (sumTwo < target):
                while (nums[pLeft]==valLeft and pLeft < pRight):
                    pLeft+=1
            elif (sumTwo > target):
                while (nums[pRight]==valRight and pLeft < pRight):
                    pRight-=1
        return result
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        result = []
        isAppended = False
        appendNum = None
        for idx, num in enumerate(nums):

            if (isAppended):
                if (num == appendNum):
                    continue
                else:
                    isAppended = False
                    appendNum = None

            twoSumsPairs = self.sortedTwoSumPairs(nums, idx+1, 0-num)

            for pair in twoSumsPairs:
                result.append([num, pair[0], pair[1]])
            isAppended = True
            appendNum = num

        return result



            


        
        
# @lc code=end

File: LeetCode_Go/List&LinkedList/15.3-sum.go
/*
 * @lc app=leetcode id=15 lang=golang
 *
 * [15] 3Sum
 */

// @lc code=start

// Given an integer array nums, return all the non-duplicate triplets [nums[i], nums[j], nums[k]] such that i != j, i != k, and j != k, and nums[i] + nums[j] + nums[k] == 0.

// Sort the input array. Implement `2Sum(nums, n)` to find unique twoSums that sums to `n` in a sorted array. Iterate the array and for each `first encounter of a unique element` nums[i] call 2Sum(nums[i+1:], 0-nums[i])

func threeSum(nums []int) [][]int {
	quickSort(nums, 0, len(nums)-1)
	var result [][]int
	i := 0
	for i < len(nums) {
		// find twoSums
		twoSumResult := twoSum(nums, i+1, 0-nums[i])
		if len(twoSumResult) > 0 {
			for idx := range twoSumResult {
				twoSumResult[idx] = append(twoSumResult[idx], nums[i])
			}
			result = append(result, twoSumResult...)
		}
		// skip all repeated values
		val := nums[i]
		for i < len(nums) && nums[i] == val {
			i++
		}
	}
	return result
}

// find unique pairs that sums to number n in sorted array nums[i:]
func twoSum(nums []int, i, n int) [][]int {
	var result [][]int
	i, j := i, len(nums)-1
	for i < j {
		sum := nums[i] + nums[j]
		if sum < n {
			i++
		} else if sum > n {
			j--
		}
		if sum == n {
			result = append(result, []int{nums[i], nums[j]})
			// skip all repeated elements
			leftVal, rightVal := nums[i], nums[j]
			for i < j && nums[i] == leftVal {
				i++
			}
			for i < j && nums[j] == rightVal {
				j--
			}
		}
	}
	return result
}

func quickSort(nums []int, i, j int) {
	// base case: empty or single element array
	if i == j || i > j {
		return
	}

	// pre recursion
	target := nums[i]
	pSmall := i + 1
	pEdit := i + 1
	for pSmall <= j {
		// if find one smaller than target, sawp pEdit and pSmall, and move pEdit.
		if nums[pSmall] < target {
			nums[pSmall], nums[pEdit] = nums[pEdit], nums[pSmall]
			pEdit++
		}
		pSmall++
	}

	// swap pEdit - 1 and target
	nums[i], nums[pEdit-1] = nums[pEdit-1], nums[i]

	// recursion
	quickSort(nums, i, pEdit-2)
	quickSort(nums, pEdit, j)
}

// @lc code=end

